worker/src/services/IngestionService/index.ts

```diff
@@ -0,0 +1,788 @@
+        projectId,
+      });
+      const serializedEventData = JSON.stringify({ ...event, projectId });
+
+      await this.redis.lpush(projectEntityKey, serializedEventData);
+      await this.redis.expire(projectEntityKey, this.bufferTtlSeconds);
+      await this.ingestionFlushQueue.add(QueueJobs.FlushIngestionEntity, null, {
+        jobId: projectEntityKey,
+      });
+    });
+
+    await Promise.all(ingestedEvents);
+  }
+
+  public async flush(projectEntityKey: string): Promise<void> {
+    const eventList = (await this.redis.lrange(projectEntityKey, 0, -1))
+      .map((serializedEventData) => {
+        const parsed = ingestionEventWithProjectId.safeParse(
+          JSON.parse(serializedEventData)
+        );
+
+        if (!parsed.success) {
+          logger.error(
+            `Failed to parse event ${serializedEventData} : ${parsed.error}`
+          );
+
+          return null;
+        }
+
+        return parsed.data;
+      })
+      .filter(Boolean) as IngestionEventWithProjectIdType[];
+
+    if (eventList.length === 0) {
+      throw new Error(
+        `No valid events found in buffer for project entity ${projectEntityKey}`
+      );
+    }
+
+    const { projectId, eventType, entityId } =
+      this.parseProjectEntityKey(projectEntityKey);
+
+    switch (eventType) {
+      case EntityType.Trace:
+        return await this.processTraceEventList({
+          projectId,
+          entityId,
+          traceEventList: eventList as TraceEventType[],
+        });
+      case EntityType.Observation:
+        return await this.processObservationEventList({
+          projectId,
+          entityId,
+          observationEventList: eventList as ObservationEvent[],
+        });
+      case EntityType.Score: {
+        return await this.processScoreEventList({
+          projectId,
+          entityId,
+          scoreEventList: eventList as ScoreEventType[],
+        });
+      }
+    }
+  }
+
+  private getEventType(event: IngestionEventType): EntityType {
+    switch (event.type) {
+      case eventTypes.TRACE_CREATE:
+        return EntityType.Trace;
+      case eventTypes.OBSERVATION_CREATE:
+      case eventTypes.OBSERVATION_UPDATE:
+      case eventTypes.EVENT_CREATE:
+      case eventTypes.SPAN_CREATE:
+      case eventTypes.SPAN_UPDATE:
+      case eventTypes.GENERATION_CREATE:
+      case eventTypes.GENERATION_UPDATE:
+        return EntityType.Observation;
+      case eventTypes.SCORE_CREATE:
+        return EntityType.Score;
+      case eventTypes.SDK_LOG:
+        return EntityType.SDK_LOG;
+    }
+  }
+
+  private getProjectEntityKey(params: {
+    projectId: string;
+    eventType: EntityType;
+    entityId: string;
+  }): string {
+    return `ingestionBuffer_${params.projectId}_${params.eventType}_${params.entityId}`;
+  }
+
+  private parseProjectEntityKey(projectEntityKey: string) {
+    const split = projectEntityKey.split("_");
+
+    if (split.length !== 4) {
+      throw new Error(
```

Comment:
entitykey can be set by our users. ProjectId and eventType cannot. Hence i would only check for the first three "_". Everything after that is user generated.